### PR TITLE
Fixes #24743 - medium provider template check works

### PR DESCRIPTION
**Pull request by lzap**
Click on a Build button and the error appears, for me on Kickstart default iPXE template - all the others are valid for some reason. Full stack:

```
Rendering Kickstart default iPXE
Review template error
 | Foreman::Exception: ERF42-1709 [Foreman::Exception]: Must supply an entity to find a medium provider
 | /home/lzap/work/foreman/app/registries/foreman/plugin/medium_providers_registry.rb:23:in `find_provider'
 | /home/lzap/work/foreman/app/models/concerns/host_template_helpers.rb:38:in `medium_provider'
 | /home/lzap/work/foreman/app/models/concerns/host_common.rb:83:in `url_for_boot'
 | /home/lzap/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/safemode-1.3.5/lib/safemode/jail.rb:31:in `method_missing'
 | Kickstart default iPXE:69:in `bind'
 | /home/lzap/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/safemode-1.3.5/lib/safemode.rb:51:in `eval'
 | /home/lzap/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/safemode-1.3.5/lib/safemode.rb:51:in `eval'
 | /home/lzap/work/foreman/lib/foreman/renderer/safe_mode_renderer.rb:7:in `render'
 | /home/lzap/work/foreman/lib/foreman/renderer/base_renderer.rb:16:in `render'
 | /home/lzap/work/foreman/lib/foreman/renderer.rb:45:in `render'
 | /home/lzap/work/foreman/app/models/template.rb:153:in `render'
 | /home/lzap/work/foreman/app/models/host/base.rb:351:in `render_template'
 | /home/lzap/work/foreman/app/services/host_build_status.rb:36:in `block in templates_status'
 | /home/lzap/work/foreman/app/services/host_build_status.rb:33:in `each'
 | /home/lzap/work/foreman/app/services/host_build_status.rb:33:in `templates_status'
 | /home/lzap/work/foreman/app/services/host_build_status.rb:15:in `check_all_statuses'
 | /home/lzap/work/foreman/app/models/host/managed.rb:733:in `build_status_checker'
 | /home/lzap/work/foreman/app/controllers/hosts_controller.rb:226:in `review_before_build'
```

### CODE REVIEW

File: app/models/concerns/host_template_helpers.rb [hunk 1]
```diff
@@ -35,6 +35,6 @@ def url_options
   end
 
   def medium_provider
-    @medium_provider ||= Foreman::Plugin.medium_providers.find_provider(@host)
+    @medium_provider ||= Foreman::Plugin.medium_providers.find_provider(@host || self)
```
Comment (timogoebel) [inline app/models/concerns/host_template_helpers.rb]: In what case is `@host` present? If the comment on top of this file is correct, this is mixed into host. So `self` should be the only valid option.

**Review by timogoebel — CHANGES_REQUESTED** [review]
@lzap: Thanks, do you mind adding a small unit test to cover this?

```
assert_equal 'whatever', @host.medium_provider 
```

**Comment by ShimShtein** [comment]
We already have [`host_medium_provider`](https://github.com/theforeman/foreman/blob/ef1bbc264acfe0864bd24e7ce54e6b37dcda37ad/app/models/concerns/host_common.rb#L77) that does accesses `self`.

The reason for both `host_medium_provider` and `medium_provider` is that in some cases, both helpers are added, and then they overload each other.

**Comment by timogoebel** [comment]
The way I see it,

`HostTemplateHelpers.medium_provider` is added to `Host`
`HostCommon.host_medium_provider` is added to `Host` and `Hostgroup`

However both methods are basically identical. The latter however passes `@host`, but I don't think this is defined anymore after the templates refactoring (it was `@host = self` before iirc). It should always be `self`. So both methods are identical. Do we need both of them?

**Comment by ShimShtein** [comment]
> Do we need both of them?

@timogoebel, I would be glad to get rid of one of them. As long as the templates would render, and tests will pass of course :smiley:

**Comment by timogoebel** [comment]
@ShimShtein: I'd vote for the removal of `host_medium_provider` and moving `medium_provider` to `HostCommon `. :-)

So the end goal would be to have `@host.medium_provider` and `@hostgroup.medium_provider`.

**Comment by ShimShtein** [comment]
@timogoebel  :+1: from me.

@lzap, would you do the honours, since you have started it :wink: ?

**Comment by lzap** [comment]
Ok done major tests are passing locally, let's see what our CI tells me.

**Comment by ShimShtein** [comment]
Could you please add an assertion for the value? I mean that the value is full URL path leading to expected location? There is a caveat about parameter resolving in the URL ($arch e.t.c.), it would be nice if the test actually made sure that the parameters are interpolated correctly.

**Comment by lzap** [comment]
@ShimShtein I don't understand, I am asserting media provider presence, this patch/test is not about functionality of medium provider. I can add something like this but unsure about it:

```ruby
  test "host have a media provider providing a valid URL" do
    hostgroup = FactoryBot.create(:hostgroup, :with_domain, :with_os)
    host = FactoryBot.create(:host, :managed, :hostgroup => hostgroup)
    assert_match /^http:/, host.medium_provider.medium_uri.to_s
  end
```

**Comment by ShimShtein** [comment]
@lzap I think it's not relevant anymore. 
:+1: ACK from me.

**Comment by lzap** [comment]
@timogoebel a test is present, this is good to go.

**Review by timogoebel — DISMISSED** [review]
LGTM.

**Comment by timogoebel** [comment]
@lzap: This breaks Katello. Do you want to file a PR to fix this there as well?

**Comment by mmoll** [comment]
@lzap this needs a rebase :)

**Comment by lzap** [comment]
Rebased, created PR for katello.

**Comment by lzap** [comment]
This is kinda blocker for all provisioning testing, @timogoebel @mmoll @ShimShtein can you take a quick look? Tests are green, katello PR pending. TY

**Review by timogoebel — APPROVED** [review]
LGTM, I'll merge.

**Comment by timogoebel** [comment]
Thanks, @lzap.

**Pull request merged**